### ISSUE TITLE
Saving an entry is very slow when Scout is enabled

### ISSUE THREAD

**Issue by martinhellwagner** [studioespresso/craft-scout#223]
We're running a large Craft install with close to 100 multisites and dozens of plugins. However, since we're running it on good hardware, saving is pretty fast and smooth – until we enable the Scout plugin. Is there anything that Scout does when saving an entry that takes such a long time, like indexing the whole page? If so, is it possible to do that with a Cron Job, like refreshing the indexes with `php craft scout/index/refresh`?

**Comment by martinhellwagner** [comment]
@janhenckens 
@timkelty 

Any ETA on the PR from your side? The saving process on our Production site is unfortunately very slow because of this, and also leads to slugs not being updated (see https://github.com/craftcms/cms/issues/9888).

**Comment by martinhellwagner** [comment]
> @martinhellwagner do you have a dedicated queue listener for Craft, does your Scout config have `queue => true`?

Yes, we're running the queue with `php craft queue/listen`.

**Comment by martinhellwagner** [comment]
@janhenckens 
@timkelty 

Any ETA on a PR for this issue? Even with the parameter `queue => true`, saving takes quite a long time. We're running our queue with the command `php craft queue/listen`.

**Comment by timkelty** [comment]
@martinhellwagner would you be able to send over a composer.json/lock and a db dump tp support@craftcms.com so I could try and diagnose locally?

**Comment by martinhellwagner** [comment]
@timkelty 

I did that already. I've been talking to Brad Bell about it – my request number is 16592.

**Comment by janhenckens** [comment]
Disabling automatic relation-indexing might be worth a try here too (`'indexRelations' => false`, in the latest version of Scout)

**Comment by martinhellwagner** [comment]
@janhenckens 

I've just tried that – unfortunately, saving is still slow. Only disabling Scout makes entry saving fast.

**Comment by martinhellwagner** [comment]
@timkelty 
@janhenckens 

I'm wondering if Scout update are really pushed into the queue? Even with Scout enabled, I can't see any extra queue jobs being pushed. Or is nothing pushed at all when `'sync' => false` is set? In that case however, it doesn't make sense that saving is so much slower when Scout is enabled.

**Comment by martinhellwagner** [comment]
One more thought: we're running our queue in Redis – can that be a problem? When switching to the Craft Queue, saving seems fast even with Scout enabled.

**Comment by janhenckens** [comment]
@martinhellwagner Are your other queue jobs running is they should? 

When `sync` is set to false, Scout should **not** be doing anything on entry save. 

Could you maybe send me the composer lock & json + a db dump as well (jan at henckens.be) so I can have a look? Thanks!

**Comment by martinhellwagner** [comment]
@janhenckens 

Yes, other queue jobs are running normally. If Scout is not doing anything on entry save, how come saving is still much slower when Scout is enabled? Is there a chance this parameter is not read correctly?

I've sent over the Composer files to @timkelty about a week ago – will you be able to access them, or should I send them anew?

**Comment by janhenckens** [comment]
@martinhellwagner Could you send them over again? Then I can have a look right away.

**Comment by martinhellwagner** [comment]
@janhenckens 

I'm just in the process of uploading the Database to Google Drive – it's pretty big (5 GB). Will update this message when the upload is finished.

Here's the link to the zipped file containing the composer files as well as the config folder: https://drive.google.com/file/d/1WHZLLFEQk7tXyMoRWeH434vPSjcMttvw/view?usp=sharing

**Comment by timkelty** [comment]
@martinhellwagner can you tell us a bit about your redis instance and connection looks like? I wonder if there is some throttling going on there…

**Comment by martinhellwagner** [comment]
@timkelty 

It's really just a simple Redis instance – all the configuration is in the `app.php` config file:

```
        'components' => [
            // Cache schema
            'db' => function() {
                $config = craft\helpers\App::dbConfig();
                $config['enableSchemaCache'] = true;
                $config['schemaCacheDuration'] = 60 * 60 * 24; // 1 day
                return Craft::createObject($config);
            },

            // Initialize Redis component
            'redis' => [
                'class' => yii\redis\Connection::class,
                'hostname' => App::env('REDIS_HOSTNAME'),
                'port' => App::env('REDIS_PORT'),
                'database' => App::env('REDIS_DEFAULT_DB'),
                'password' => null,
            ],

            // Use cache with Redis
            'cache' => [
                'class' => yii\redis\Cache::class,
                'keyPrefix' => '_' . App::env('APP_ID'). '_CACHE_' ?: 'CraftCMS',
                'defaultDuration' => 86400,
                'redis' => 'redis',
            ],

            // Use queue with Redis
            'queue' => [
                'class' => yii\queue\redis\Queue::class,
                'redis' => 'redis',
                'channel' => 'queue',
            ],
        ],
```

**Comment by timkelty** [comment]
@martinhellwagner right – what I'm more interested in is the actual instance itself (where it is hosted, specs, etc).
You could be experiencing some network issues, if the redis host is a remote connection.

**Comment by martinhellwagner** [comment]
Nope, it's on the same server as the MySQL DB and the code. The server has pretty large specs and is not experiencing any CPU/RAM shortages whatsoever.

**Comment by janhenckens** [comment]
@martinhellwagner I have the site up & running, could you point me to an entry/site that is having the issue or is it on just about every save/entry?

**Comment by martinhellwagner** [comment]
@janhenckens 

It's very random to be honest, and could happen to any entry – mostly to the entries with lots of relations though. The one I was always testing with was `/admin/entries/pages/30381-zusammenarbeit?site=default` – if you change the slug for example, saving takes a long time, and probably because of that the slug sometimes doesn't get updated.

**Comment by markhuot** [comment]
This line is the culprit (in my testing). The new `indexRelations` seemed to help. However, if you have `maxRevisions` set then every save is also a delete and that triggers the same massive query for related elements.

https://github.com/studioespresso/craft-scout/blob/cc68a24e1524fd3f63f5f5e64f615a8c644673a5/src/Scout.php#L166

Is it possible to throw `beforeDeleteRelated` behind the same `indexRelations` flag as well for larger datasets?

**Comment by timkelty** [comment]
@markhuot that sounds like a great idea!

**Comment by janhenckens** [comment]
The fix @markhuot mentioned is out in 2.6.1.

**Comment by martyspain** [comment]
This is a game-changing fix for us. I was having nightmares trying to figure out why saving was taking so long on our Craft install and throwing ever-more expensive and overpowered hardware at the problem was barely scratching the surface. Just implementing `'indexRelations' => false` in our config and updating to 2.6.1 has dropped our save times from over 3 minutes to just a few seconds. Thank you all for getting to the bottom of this!
